### PR TITLE
Add fed IDP_ID column to IDN_FED_AUTH_SESSION_MAPPING table

### PR DESCRIPTION
**Pull request by ThumulaPerera**
## Proposed Changes in this PR

- Fixes : https://github.com/wso2/product-is/issues/15649 (NPE while processing SLO request from federated IDP when the same IDP is configured twice.)
- Introduces a new column called IDP_ID in IDN_FED_AUTH_SESSION_MAPPING table to keep the IDP ID of the federated IDP.
- The changes described below will be activated only if the IDP_ID column exists in IDN_FED_AUTH_SESSION_MAPPING table.
- Adds the IDP ID when a federated auth session is added to the table and updated in the table.
- When removing an entry, IDP ID will be used as well.
- In the case of a logout request initiated by the federated IDP side, is sessionContext is not set, a special parameter FED IDP ID sent by outbound SAML authenticator is used to determine the IDP_ID of the entry to be removed.

## Related PRs

Outbound SAML fix to retrieve the session details from the IDN_FED_AUTH_SESSION_MAPPING table [1].
IDN_FED_AUTH_SESSION_MAPPING table column update PR [2].

## When should this PR be merged
This PR should be merged before merging [1].

[1] https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/161
[2] https://github.com/wso2/carbon-identity-framework/pull/4559

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 1]
```diff
@@ -651,6 +644,73 @@ protected void concludeFlow(HttpServletRequest request, HttpServletResponse resp
...
+
+        UserSessionStore userSessionStore = UserSessionStore.getInstance();
+        if (!userSessionStore.hasExistingFederatedAuthSession(
+                authHistory.getIdpSessionIndex())) {
+            userSessionStore.storeFederatedAuthSessionInfo(sessionContextKey, authHistory);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Federated auth session with the id: %s already " +
+                        "exists.", authHistory.getIdpSessionIndex()));
+            }
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:658]: Is there any need to use String.format given that only one param is passed? We can simply concat

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 2]
```diff
@@ -185,13 +185,23 @@ public class SQLQueries {
...
+            + "(IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, PROTOCOL_TYPE) VALUES (?, ?, ?, ?, ?)";
 
     // Store federated authentication session details to map the session context key with the idp session index.
     public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO_WITH_TENANT = "INSERT INTO " +
             "IDN_FED_AUTH_SESSION_MAPPING (IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, " +
             "PROTOCOL_TYPE, TENANT_ID) VALUES (?, ?, ?, ?, ?, ?)";
 
+    // Store federated authentication session details to map the session context key with the idp session index.
+    public static final String SQL_STORE_FEDERATED_AUTH_SESSION_INFO_WITH_IDP_ID = "INSERT INTO " +
+            "IDN_FED_AUTH_SESSION_MAPPING (IDP_SESSION_ID, SESSION_ID, IDP_NAME,  AUTHENTICATOR_ID, " +
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:197]:
```suggestion
            "IDN_FED_AUTH_SESSION_MAPPING (IDP_SESSION_ID, SESSION_ID, IDP_NAME, AUTHENTICATOR_ID, " +
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java (hunk 1 above)
Comment (ThumulaPerera) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java]: Fixed with [f22af54](https://github.com/wso2/carbon-identity-framework/pull/4558/commits/f22af54d48148c23a43fda73c6b52870762a599c)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java (hunk 2 above)
Comment (ThumulaPerera) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java]: Fixed with [f22af54](https://github.com/wso2/carbon-identity-framework/pull/4558/commits/f22af54d48148c23a43fda73c6b52870762a599c)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java [hunk 3]
```diff
@@ -200,13 +210,29 @@ public class SQLQueries {
     public static final String SQL_GET_FEDERATED_AUTH_SESSION_ID_BY_SESSION_ID_WITH_TENANT = "SELECT SESSION_ID " +
             "FROM IDN_FED_AUTH_SESSION_MAPPING WHERE IDP_SESSION_ID = ? AND TENANT_ID = ?";
 
+    // Get federated authentication session id using the idp session id and the idp id.
+    public static final String SQL_GET_FEDERATED_AUTH_SESSION_ID_BY_SESSION_ID_WITH_IDP_ID = "SELECT SESSION_ID " +
+            "FROM IDN_FED_AUTH_SESSION_MAPPING WHERE IDP_SESSION_ID = ? AND IDP_ID = ?";
+
+    // Get federated authentication session id using the idp session id.
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java:217]:
```suggestion
    // Get federated authentication session id using the tenant id and IdP id.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 4]
```diff
@@ -890,6 +890,67 @@ public void storeFederatedAuthSessionInfo(String sessionContextKey, AuthHistory
...
+            prepStmt.setInt(7, idpId);
+            prepStmt.execute();
+        } catch (SQLException e) {
+            String msg = String.format("Error while adding session details of the session index: %s, IdP: %s " +
+                    "and tenant id: %s.", sessionContextKey, authHistory.getIdpName(), tenantId);
+            throw new UserSessionException(msg, e);
+        }
+    }
+
+
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:953]:
```suggestion
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java [hunk 5]
```diff
@@ -975,7 +1094,7 @@ public boolean hasExistingFederatedAuthSession(String idpSessionIndex) throws Us
      * @return True if a federated auth session found with the given session index.
      * @throws UserSessionException If an error occurred while checking for an federated auth session.
      */
-    public boolean isExistingFederatedAuthSessionAvailable(String idpSessionIndex, int tenantId)
+    public boolean hasExistingFederatedAuthSession(String idpSessionIndex, int tenantId)
```
Comment (sachiniWettasinghe) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java:1097]: why have we changed a public method name here?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java (hunk 3 above)
Comment (ThumulaPerera) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SQLQueries.java]: Fixed with [1c921fe](https://github.com/wso2/carbon-identity-framework/pull/4558/commits/1c921fee7a66cc83e784b863c8fae6a4e13e902e)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java (hunk 4 above)
Comment (ThumulaPerera) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java]: Fixed with [1c921fe](https://github.com/wso2/carbon-identity-framework/pull/4558/commits/1c921fee7a66cc83e784b863c8fae6a4e13e902e)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java (hunk 5 above)
Comment (ThumulaPerera) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/UserSessionStore.java]: Reverted to back to the previous name with [1c921fe](https://github.com/wso2/carbon-identity-framework/pull/4558/commits/1c921fee7a66cc83e784b863c8fae6a4e13e902e)

**Review by sachiniWettasinghe — APPROVED** [review]

**Pull request merged**